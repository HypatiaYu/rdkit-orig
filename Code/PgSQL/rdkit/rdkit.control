comment = 'Cheminformatics functionality for PostgreSQL.'
default_version = '3.1'
module_pathname = '$libdir/rdkit'
relocatable = true