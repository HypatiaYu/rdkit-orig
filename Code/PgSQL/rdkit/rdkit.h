// $Id$
//
//  Copyright (c) 2010, Novartis Institutes for BioMedical Research Inc.
//  All rights reserved.
// 
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met: 
//
//     * Redistributions of source code must retain the above copyright 
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
//       copyright notice, this list of conditions and the following 
//       disclaimer in the documentation and/or other materials provided 
//       with the distribution.
//     * Neither the name of Novartis Institutes for BioMedical Research Inc. 
//       nor the names of its contributors may be used to endorse or promote 
//       products derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
#ifndef _RDKIT_H_
#define _RDKIT_H_

#ifdef __cplusplus
extern "C" {
#endif

#include "postgres.h"

typedef bytea Mol;

#define DatumGetMolP(x)		((Mol*)PG_DETOAST_DATUM(x))
#define DatumGetMolPCopy(x)	((Mol*)PG_DETOAST_DATUM_COPY(x))
#define MolPGetDatum(x)		(PointerGetDatum(x))

#define PG_GETARG_MOL_P(x) 	DatumGetMolP(PG_GETARG_DATUM(x))
#define PG_GETARG_MOL_P_COPY(x) DatumGetMolPCopy(PG_GETARG_DATUM(x))
#define PG_RETURN_MOL_P(x)	PG_RETURN_DATUM(MolPGetDatum(x))

typedef bytea BitmapFingerPrint;

#define DatumGetBitmapFingerPrintP(x)		((BitmapFingerPrint*)PG_DETOAST_DATUM(x))
#define DatumGetBitmapFingerPrintPCopy(x)	((BitmapFingerPrint*)PG_DETOAST_DATUM_COPY(x))
#define BitmapFingerPrintPGetDatum(x)		(PointerGetDatum(x))

#define PG_GETARG_BITMAPFINGERPRINT_P(x) 	DatumGetBitmapFingerPrintP(PG_GETARG_DATUM(x))
#define PG_GETARG_BITMAPFINGERPRINT_P_COPY(x) DatumGetBitmapFingerPrintPCopy(PG_GETARG_DATUM(x))
#define PG_RETURN_BITMAPFINGERPRINT_P(x)	PG_RETURN_DATUM(BitmapFingerPrintPGetDatum(x))

typedef bytea SparseFingerPrint;

#define DatumGetSparseFingerPrintP(x)		((SparseFingerPrint*)PG_DETOAST_DATUM(x))
#define DatumGetSparseFingerPrintPCopy(x)	((SparseFingerPrint*)PG_DETOAST_DATUM_COPY(x))
#define SparseFingerPrintPGetDatum(x)		(PointerGetDatum(x))

#define PG_GETARG_SPARSEFINGERPRINT_P(x) 	DatumGetSparseFingerPrintP(PG_GETARG_DATUM(x))
#define PG_GETARG_SPARSEFINGERPRINT_P_COPY(x) DatumGetSparseFingerPrintPCopy(PG_GETARG_DATUM(x))
#define PG_RETURN_SPARSEFINGERPRINT_P(x)	PG_RETURN_DATUM(SparseFingerPrintPGetDatum(x))

/*
 * GUC
 */
extern double getTanimotoLimit(void);
extern double getDiceLimit(void);

/*
 * From/to C/C++
 */

/* RDKit::ROMol */
typedef void * CROMol; 
void	freeCROMol(CROMol data);

CROMol constructROMol(Mol* data); 
Mol * deconstructROMol(CROMol data); 

CROMol parseMolText(char *data,bool asSmarts);
char *makeMolText(CROMol data, int *len,bool asSmarts);
bool isValidSmiles(char *data);
bool isValidSmarts(char *data);

int molcmp(CROMol i, CROMol a);

int MolSubstruct(CROMol i, CROMol a);

bytea *makeMolSign(CROMol data);

double MolAMW(CROMol i);
double MolLogP(CROMol i);
int MolHBA(CROMol i);
int MolHBD(CROMol i);
int MolNumAtoms(CROMol i);
int MolNumHeavyAtoms(CROMol i);


/* ExplicitBitVect */
typedef void * MolBitmapFingerPrint;
void	freeMolBitmapFingerPrint(MolBitmapFingerPrint data);

MolBitmapFingerPrint constructMolBitmapFingerPrint(BitmapFingerPrint *data);
BitmapFingerPrint * deconstructMolBitmapFingerPrint(MolBitmapFingerPrint data);
bytea * makeSignatureBitmapFingerPrint(MolBitmapFingerPrint data);

int MolBitmapFingerPrintSize(MolBitmapFingerPrint a);

double calcBitmapTanimotoSml(MolBitmapFingerPrint a, MolBitmapFingerPrint b);
double calcBitmapDiceSml(MolBitmapFingerPrint a, MolBitmapFingerPrint b);

/* SparseIntVect<boost::int32_t> */
typedef void * MolSparseFingerPrint;
void	freeMolSparseFingerPrint(MolSparseFingerPrint data);

MolSparseFingerPrint constructMolSparseFingerPrint(SparseFingerPrint *data);
SparseFingerPrint * deconstructMolSparseFingerPrint(MolSparseFingerPrint data);
bytea * makeSignatureSparseFingerPrint(MolSparseFingerPrint data, int numBits);
bytea * makeLowSparseFingerPrint(MolSparseFingerPrint data, int numInts);

double calcSparseTanimotoSml(MolSparseFingerPrint a, MolSparseFingerPrint b);
double calcSparseDiceSml(MolSparseFingerPrint a, MolSparseFingerPrint b);
double calcSparseStringDiceSml(const char *a, unsigned int sza, const char *b, unsigned int szb);
MolSparseFingerPrint  addSFP(MolSparseFingerPrint a, MolSparseFingerPrint b);
MolSparseFingerPrint  subtractSFP(MolSparseFingerPrint a, MolSparseFingerPrint b);


void countOverlapValues(bytea * sign, MolSparseFingerPrint data, int numBits,
        int * sum, int * overlapSum, int * overlapN);
void countLowOverlapValues(bytea * sign, MolSparseFingerPrint data, int numInts,
        int * querySum, int *keySum, int * overlapUp, int * overlapDown);
/*
 * Various mol -> fp transformation
 */

MolBitmapFingerPrint makeLayeredBFP(CROMol data);
MolBitmapFingerPrint makeRDKitBFP(CROMol data);
MolBitmapFingerPrint makeMorganBFP(CROMol data, int radius);
MolSparseFingerPrint makeMorganSFP(CROMol data, int radius);
MolSparseFingerPrint makeAtomPairSFP(CROMol data);
MolSparseFingerPrint makeTopologicalTorsionSFP(CROMol data);
MolBitmapFingerPrint makeAtomPairBFP(CROMol data);
MolBitmapFingerPrint makeTopologicalTorsionBFP(CROMol data);

/*
 * Indexes
 */

#define NUMBITS					(2048)
#define NUMRANGE				(120)

#define INTRANGEMAX				(0xff)
typedef struct IntRange {
	uint8		low;
	uint8		high;
} IntRange;

#define RDKitTanimotoStrategy   (1)
#define RDKitDiceStrategy       (2)
#define	RDKitContains			(3)
#define	RDKitContained			(4)
#define	RDKitEquals			(6)

bool calcConsistency(bool isLeaf, uint16 strategy, 
    double nCommonUp, double nCommonDown, double nKey, double nQuery);


/*
 *  Cache subsystem. Moleculas and fingerprints I/O is extremely expensive.
 */
struct MemoryContextData; /* forward declaration to prevent conflicts with C++ */
void* SearchMolCache( void *cache, struct MemoryContextData * ctx, Datum a, 
						Mol **m, CROMol *mol, bytea **sign);
void* SearchBitmapFPCache( void *cache, struct MemoryContextData * ctx, Datum a, 
							BitmapFingerPrint **f, MolBitmapFingerPrint *fp, bytea **val);
void* SearchSparseFPCache( void *cache, struct MemoryContextData * ctx, Datum a, 
							SparseFingerPrint **f, MolSparseFingerPrint *fp, bytea **val);

/*
void* SearchMolCache( void *cache, struct MemoryContextData * ctx, Datum a, Mol **m, CROMol *mol, MolFingerPrint *fp, bytea **val);
void* SearchFPCache( void *cache, struct MemoryContextData * ctx, Datum a, FingerPrint **f, MolFingerPrint *fp, bytea **val);
*/
#ifdef __cplusplus
}
#endif
#endif
