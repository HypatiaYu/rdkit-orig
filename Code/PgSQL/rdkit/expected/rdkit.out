--
-- first, define the datatype.  Turn off echoing so that expected file
-- does not depend on contents of rdkit.sql.
--
SET client_min_messages = warning;
\set ECHO none
RESET client_min_messages;
SELECT is_valid_smiles('c1ccccc1');
 is_valid_smiles 
-----------------
 t
(1 row)

SELECT is_valid_smiles('c1ccccc');
 is_valid_smiles 
-----------------
 f
(1 row)

SELECT is_valid_smiles('c1cccn1');
 is_valid_smiles 
-----------------
 f
(1 row)

CREATE TABLE pgmol (id int, m mol);
\copy pgmol from 'data/data'
CREATE UNIQUE INDEX mol_ididx ON pgmol (id);
SELECT count(*) FROM pgmol;
 count 
-------
  1000
(1 row)

SELECT count(*) FROM pgmol WHERE m @> 'c1ccccc1';
 count 
-------
   901
(1 row)

SELECT count(*) FROM pgmol WHERE m @> 'c1cccnc1';
 count 
-------
   245
(1 row)

SELECT count(*) FROM pgmol WHERE 'c1ccccc1' <@ m;
 count 
-------
   901
(1 row)

SELECT count(*) FROM pgmol WHERE 'c1cccnc1' <@ m;
 count 
-------
   245
(1 row)

SELECT id, layered_fp(m) AS f INTO pgbfp FROM pgmol;
CREATE UNIQUE INDEX bfp_ididx ON pgbfp (id);
SELECT id, morgan_fp(m,1) AS f INTO pgsfp FROM pgmol;
CREATE UNIQUE INDEX sfp_ididx ON pgsfp (id);
SELECT id, rdkit_fp(m) AS f INTO pgavfp FROM pgmol;
SELECT id, torsion_fp(m) AS f INTO pgtorsfp FROM pgmol;
SELECT id, atompair_fp(m) AS f INTO pgpairfp FROM pgmol;
set rdkit.tanimoto_threshold=0.5;
set rdkit.dice_threshold=0.5;
SELECT 
	id, 
	tanimoto_sml(layered_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol), f) 
FROM
	 (SELECT * FROM pgbfp ORDER BY id) AS t 
WHERE layered_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol) % f  
LIMIT 10;
   id   |   tanimoto_sml    
--------+-------------------
   2722 |  0.51394422310757
  11829 | 0.557377049180328
  63248 | 0.606870229007634
  66722 | 0.734939759036145
  68281 |             0.604
  98960 | 0.521311475409836
 198995 | 0.611570247933884
 230488 | 0.657142857142857
 233331 | 0.534883720930233
 246547 | 0.548951048951049
(10 rows)

SELECT 
	id, 
	dice_sml(layered_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol), f) 
FROM
	 (SELECT * FROM pgbfp ORDER BY id) AS t 
WHERE layered_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol) % f  
LIMIT 10;
   id   |     dice_sml      
--------+-------------------
   2722 | 0.678947368421053
  11829 | 0.715789473684211
  63248 | 0.755344418052256
  66722 | 0.847222222222222
  68281 | 0.753117206982544
  98960 | 0.685344827586207
 198995 | 0.758974358974359
 230488 | 0.793103448275862
 233331 | 0.696969696969697
 246547 | 0.708803611738149
(10 rows)

SELECT 
	id, 
	tanimoto_sml(layered_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol), f) 
FROM
	 (SELECT * FROM pgbfp ORDER BY id) AS t 
WHERE layered_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol) # f  
LIMIT 10;
  id   |   tanimoto_sml    
-------+-------------------
  2722 |  0.51394422310757
  2788 | 0.475836431226766
 11829 | 0.557377049180328
 42640 | 0.471810089020772
 63248 | 0.606870229007634
 66722 | 0.734939759036145
 68281 |             0.604
 75846 | 0.448132780082988
 82533 | 0.496932515337423
 83773 | 0.477341389728097
(10 rows)

SELECT 
	id, 
	dice_sml(layered_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol), f),
	size(f)
FROM
	 (SELECT * FROM pgbfp ORDER BY id) AS t 
WHERE layered_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol) # f  
LIMIT 10;
  id   |     dice_sml      | size 
-------+-------------------+------
  2722 | 0.678947368421053 | 1024
  2788 | 0.644836272040302 | 1024
 11829 | 0.715789473684211 | 1024
 42640 | 0.641129032258065 | 1024
 63248 | 0.755344418052256 | 1024
 66722 | 0.847222222222222 | 1024
 68281 | 0.753117206982544 | 1024
 75846 |   0.6189111747851 | 1024
 82533 | 0.663934426229508 | 1024
 83773 | 0.646216768916155 | 1024
(10 rows)

set rdkit.tanimoto_threshold=0.4;
SELECT 
	id, 
	tanimoto_sml(morgan_fp('C1C(OC2=CC(=CC(=C2C1=O)))'::mol, 1), f) 
FROM
	 (SELECT * FROM pgsfp ORDER BY id) AS t 
WHERE morgan_fp('C1C(OC2=CC(=CC(=C2C1=O)))'::mol, 1) % f  
LIMIT 10;
   id    |   tanimoto_sml    
---------+-------------------
 3761688 | 0.441860465116279
(1 row)

SELECT 
	id, 
	dice_sml(morgan_fp('C1C(OC2=CC(=CC(=C2C1=O)))'::mol, 1), f) 
FROM
	 (SELECT * FROM pgsfp ORDER BY id) AS t 
WHERE morgan_fp('C1C(OC2=CC(=CC(=C2C1=O)))'::mol, 1) % f  
LIMIT 10;
   id    |     dice_sml      
---------+-------------------
 3761688 | 0.612903225806452
(1 row)

SELECT 
	id, 
	tanimoto_sml(morgan_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol, 1), f) 
FROM
	 (SELECT * FROM pgsfp ORDER BY id) AS t 
WHERE morgan_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol, 1) # f  
LIMIT 10;
    id    |   tanimoto_sml    
----------+-------------------
   902176 | 0.347826086956522
  2952787 | 0.365853658536585
  5281628 | 0.346153846153846
 10560368 | 0.435897435897436
 16196768 |             0.375
(5 rows)

SELECT 
	id, 
	dice_sml(morgan_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol, 1), f)
FROM
	 (SELECT * FROM pgsfp ORDER BY id) AS t 
WHERE morgan_fp('C1C(OC2=CC(=CC(=C2C1=O)O)O)'::mol, 1) # f  
LIMIT 10;
    id    |     dice_sml      
----------+-------------------
   902176 | 0.516129032258065
  2952787 | 0.535714285714286
  5281628 | 0.514285714285714
 10560368 | 0.607142857142857
 16196768 | 0.545454545454545
(5 rows)

